library/runCmds.py

```diff
@@ -9,74 +19,81 @@
 
+# RhoCmd and its sub-classes are wrapper classes around
+# the cli cmds we run on the host machines. The
+# rho_cmd class will have five dictionaries. The
+# cmd_names dictionary maps all the commands that are associated
+# with the class to the corresponding fields(facts) requested.
+# The cmd_strings dictionary maps each command name to the
+# actual command string that is run in the cli. The results
+# so obtained are stored in the cmd_results dictionary whose
+# keys are the command names. The fields dictionary stores
+# the localized versions of the field names and the data
+# dictionary is responsible for storing all the facts collected
+# from all the classes ever invoked.
 
 
-
-
-
-# basic idea, wrapper classes around the cli cmds we run on the machines
-# to be inventories. the rho_cmd class will have a string for the
-# actually command to run, a name, a 'data' attrib storing the
-# data retrivied (probably just a dict). I think it makes the most
-# sense to go ahead and parse the results of the remote cmd when we
-# run it and store it then in the data (as opposed to doing it at
-# data read time). I think it's pretty small data sets either way, so
-# not a big deal.
-
-# If we use a dict like class for the data, we maybe able to make
-# the report generation as simple as python string formatting tricks.
-# I'd like to try to avoid type'ing the data fields and just treating
-# everything as strings, since the primary target seems to be csv
-# output.
-
 class RhoCmd(object):
-    name = "base"
-    fields = {}
 
     def __init__(self):
-        #        self.cmd_strings = cmd
-        self.cmd_results = []
+        # self.cmd_strings = cmd
+        self.cmd_results = {}
         self.data = {}
+        self.cmd_strings = {}
+        self.cmd_names = {}
+        self.name = "cmd_name"
+        self.fields = {}
 
-    # we're not actually running the class on the hosts, so
-    # we will need to populate it with the output of the ssh stuff
-    # we can send a list of commands, so we expect output to be a list
-    # of output strings
+    def run_cmd(self, facts):
 
+        requested_cmd_names = []
 
+        if type(facts) == list:
+            for fact in facts:
+                for cmd_name in self.cmd_names.keys():
+                    if fact in self.cmd_names[cmd_name]:
+                        requested_cmd_names.append(cmd_name)
+
+        elif type(facts) == str:
+            if facts == 'all':
+                requested_cmd_names = self.cmd_names.keys()
+            else:
+                print "Facts not available"
+                return
+        else:
+            print "Facts not available"
+            return
+
+        for cmd_name in requested_cmd_names:
+            cmd_string = self.cmd_strings[cmd_name]
+            process = sp.Popen(cmd_string, shell=True,
+                               stdout=sp.PIPE, stderr=sp.PIPE)
+            out, err = process.communicate()
```

Comment:
What happens if this raises an exception? Do we error out of the entire command, the entire ansible run, just this particular command?